flashinfer/cute_dsl/add_rmsnorm_fp4quant.py

```diff
@@ -0,0 +1,2278 @@
+        w: torch.Tensor,
+        y: torch.Tensor,
+        s: torch.Tensor,
+        M: int,
+        eps: float,
+    ) -> None:
+        """Runtime API that converts tensors to pointers and calls the kernel."""
+        nonlocal compiled_kernel
+        compiled_kernel(
+            *get_cute_pointers([x, r, w, y, s]),
+            Int32(M),
+            Float32(eps),
+            cutlass_torch.current_stream(),
+        )
+
+    return tensor_api
+
+
+@flashinfer_api
+def add_rmsnorm_fp4quant_cute_dsl(
+    input: torch.Tensor,
+    residual: torch.Tensor,
+    weight: torch.Tensor,
+    y_fp4: torch.Tensor,
+    block_scale: torch.Tensor,
+    eps: float = 1e-6,
+    block_size: int = 16,
+    scale_format: str | None = None,
+) -> None:
+    """
+    Fused Add + RMS normalization + FP4 quantization using CuTe-DSL.
+
+    Computes: h = input + residual, y = RMSNorm(h) * weight, then quantizes to FP4.
+
+    Parameters
+    ----------
+    input : torch.Tensor
+        Input tensor, shape (batch_size, hidden_size) or (batch_size, seq_len, hidden_size).
+        Must be torch.float16 or torch.bfloat16.
+    residual : torch.Tensor
+        Residual tensor, same shape and dtype as input.
+    weight : torch.Tensor
+        Weight tensor, shape (hidden_size,). Must have the same dtype as input.
+    y_fp4 : torch.Tensor
+        Output tensor for quantized values in FP4_E2M1 format, packed as uint8.
+        Shape must be (batch_size, hidden_size // 2) or matching 3D input.
+    block_scale : torch.Tensor
+        Output tensor for block scale factors.
+        Shape must be (batch_size, hidden_size // block_size) or matching 3D input.
+    eps : float
+        Epsilon for numerical stability in RMSNorm. Default is 1e-6.
+    block_size : int
+        Block size for FP4 quantization. 16 for NVFP4, 32 for MXFP4.
+    scale_format : str, optional
+        Scale factor format: "e4m3" or "ue8m0". If None, auto-selects based on block_size.
+    """
+    is_3d = input.dim() == 3
+    if is_3d:
+        B, S, H = input.shape
+        input = input.view(B * S, H).contiguous()
+        residual = residual.view(B * S, H).contiguous()
+        y_fp4_2d = y_fp4.view(B * S, -1)
+        block_scale_2d = block_scale.view(B * S, -1)
+    else:
+        y_fp4_2d = y_fp4
+        block_scale_2d = block_scale
+
+    batch_size, hidden_size = input.shape
+    dtype = input.dtype
+
+    assert hidden_size % block_size == 0, "hidden_size must be divisible by block_size"
+    assert hidden_size >= 64, "hidden_size must be >= 64"
+    assert block_size in [16, 32], "block_size must be 16 or 32"
+
+    is_fp16 = dtype == torch.float16
+    actual_scale_format = (
+        scale_format if scale_format else ("ue8m0" if block_size == 32 else "e4m3")
+    )
+    sm_version = get_sm_version()
+
+    tensor_api = _get_compiled_kernel(
+        hidden_size, block_size, is_fp16, sm_version, actual_scale_format
+    )
+    tensor_api(
+        input.contiguous(),
```

Comment:
Do we really need input to be contiguous? I suppose we only need last dimension to be contiguous?